### PR TITLE
Introduce new `apollo-server-types` package for shared types.

### LINKED ISSUE #2398: apollo-server-plugin-base is missing dependency apollo-server-core
Similar to #1963, the `apollo-server-plugin-base` package does not declare a dependency on `apollo-server-core` in its package.json, so an error occurs when using strict package managers such as yarn pnp or pnpm, e.g.

```
ERROR in /Users/spencerelliott/Dev/path/to/node_modules/.registry.npmjs.org/apollo-server-plugin-base/0.3.4/graphql@14.1.1/node_modules/apollo-server-plugin-base/dist/index.d.ts(1,95):
TS2307: Cannot find module 'apollo-server-core/dist/requestPipelineAPI'.
```

### LINKED ISSUE #2964: apollo-server-plugin-base depends on apollo-server-core but doesn't include it as a dependency
As it was discovered here:

https://github.com/yarnpkg/berry/issues/271#issuecomment-507407938

I see in package.json apollo-server-core  is a dev dependency. is this a bug or was it intended?

### LINKED ISSUE #2970: chore: Introduce a `apollo-server-types` package.
Currently, there are a number of types which are used across many of the Apollo Server packages which are creating cycles and unnecessarily heavy dependencies on core packages for only their types.

For example, `apollo-engine-reporting` relies on (only) the `GraphQLRequestContext` type from `apollo-server-core`, but `apollo-server-core` relies heavily on `apollo-engine-reporting`.  This creates cycles in the dependency graph that we could avoid by having another package.

Likewise, packages like `apollo-server-plugin-base` currently rely on types from `apollo-server-core`, but because of a desire to maintain loose versioning between the `apollo-server-plugin-base` package and `apollo-server-core` (to maximize plugin compatibility ranges), we are currently _not_ depending on `apollo-server-core` in the way we should be in order to provide dev-time typings. (https://github.com/apollographql/apollo-server/issues/2964).

Some candidates for becoming members of a `apollo-server-types` package might be:

* `GraphQLServiceContext`
* `GraphQLRequest`
* `GraphQLResponse`
* `GraphQLRequestContext`
* `GraphQLExecutionResult`

These are just from a quick glance into `apollo-server-core/src/requestPipelineAPI`, but I'm sure there are more!

### PR DESCRIPTION
**Pull request by abernix**
This moves TypeScript types which are shared by many aspects of Apollo Server into their own, new, `apollo-server-types` package which can be depended on directly by the various parts of Apollo Server (and other Apollo tools) with minimal additional dependency weight.  This helps avoid, for example, bringing in the heavier weight of `apollo-server-core` merely to provide some type dependencies.

Furthermore, this aims to remove one of the obvious sources of cycles in our dependency graph, which occurs when `apollo-engine-reporting` depends on `apollo-server-core` (for types!), which itself depends on `apollo-engine-reporting`.  While these cycles are (roughly) fine at runtime because they're type _merely_ type dependencies, is still made abundantly obvious (read: noisy) during Lerna publishing.

As part of this project, I've also moved a couple types from `apollo-server-env` — `PromiseOrValue` and `WithRequired` — into this new `apollo-server-types` package, since they seemed less like ECMAScript compatibility needs (i.e. polyfills; which is primarily what that package aims to provide).  Providing these shared types from a package which has import side-effects just seemed like it _could be_ considered bad form and it seemed more appropriate to put them in a "types" package (now that we have it).  I've left them in `apollo-server-env` for compatibility reasons, and I suspect `apollo-server-env` will be deprecated soon and replaced with `apollo-env`.

Lastly, I've also continued to re-export these types from `packages/apollo-server-core/src/requestPipelineAPI.ts` as a matter of backwards compatibility, but after this change that single file is no longer its own TypeScript project, which was originally done to avoid cycles on `apollo-server-plugin-base`.

There's potentially more that could be added here, and it's even possible that this could be even further shared between other Apollo components (e.g. `apollo-tooling`), so feedback and input very much appreciated!

Closes #2970 
Fixes #2964
Resolves #2398

### CODE REVIEW

**Comment by JacksonKearl** [comment]
Why is `apollo-server-types` sometimes a `devDependency` and sometimes a `dependency`?

File: packages/apollo-datasource-rest/package.json [hunk 1]
```diff
@@ -22,5 +22,8 @@
     "apollo-server-env": "file:../apollo-server-env",
     "apollo-server-errors": "file:../apollo-server-errors",
     "http-cache-semantics": "^4.0.0"
+  },
+  "devDependencies": {
```
Comment (JacksonKearl) [inline packages/apollo-datasource-rest/package.json]: `devDependency`

File: packages/apollo-server-plugin-base/package.json [hunk 2]
```diff
@@ -10,8 +10,8 @@
   "engines": {
     "node": ">=6"
   },
-  "devDependencies": {
-    "apollo-server-core": "file:../apollo-server-core"
+  "dependencies": {
```
Comment (JacksonKearl) [inline packages/apollo-server-plugin-base/package.json]: `dependency`

**Comment by abernix** [comment]
@JacksonKearl 	Good question!  When I went through here, I found a couple of these were incorrect, but roughly, if the `.d.ts` type definitions emitted in the output bundle and intended for use by consuming TypeScript projects reference a type, then the package that provides that type must be listed in `dependencies`.

**Pull request merged**